File: lib-database/src/main/java/com/ustadmobile/core/db/dao/ScrapeQueueItemDao.java
package com.ustadmobile.core.db.dao;

import com.ustadmobile.lib.database.annotation.UmDao;
import com.ustadmobile.lib.database.annotation.UmQuery;
import com.ustadmobile.lib.db.entities.ScrapeQueueItem;
import com.ustadmobile.lib.db.sync.dao.BaseDao;

import java.util.List;

@UmDao
public abstract class ScrapeQueueItemDao implements BaseDao<ScrapeQueueItem> {

    public static final int STATUS_PENDING = 1;

    public static final int STATUS_RUNNING = 2;

    public static final int STATUS_DONE = 3;

    public static final int STATUS_FAILED = 4;

    public ScrapeQueueItem getNextItemAndSetStatus(int runId, int itemType) {
        ScrapeQueueItem nextItem = findNextItem(STATUS_PENDING, runId, itemType);
        if (nextItem != null) {
            updateSetStatusById(nextItem.getSqiUid(), STATUS_RUNNING);
        }

        return nextItem;
    }

    @UmQuery("SELECT * FROM ScrapeQueueItem")
    public abstract List<ScrapeQueueItem> findAll();


    @UmQuery("SELECT * FROM ScrapeQueueItem WHERE status = :status AND runId = :runId AND itemType = :itemType LIMIT 1")
    public abstract ScrapeQueueItem findNextItem(int status, int runId, int itemType);

    @UmQuery("UPDATE ScrapeQueueItem SET status = :status WHERE sqiUid = :uid")
    public abstract void updateSetStatusById(int uid, int status);

    @UmQuery("SELECT * from ScrapeQueueItem WHERE runId = :runId AND scrapeUrl = :indexUrl LIMIT 1")
    public abstract ScrapeQueueItem getExistingQueueItem(int runId, String indexUrl);

    @UmQuery("UPDATE ScrapeQueueItem SET timeStarted = :timeStarted WHERE sqiUid = :uid")
    public abstract void setTimeStarted(int uid, long timeStarted);

    @UmQuery("UPDATE ScrapeQueueItem SET timeFinished = :timeFinished WHERE sqiUid = :uid")
    public abstract void setTimeFinished(int uid, long timeFinished);

}

File: lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/dao/ScrapeQueueItemDao.kt
package com.ustadmobile.core.db.dao

import com.ustadmobile.door.annotation.DoorDao
import androidx.room.Query
import com.ustadmobile.core.db.dao.ScrapeQueueItemDaoCommon.STATUS_PENDING
import com.ustadmobile.door.lifecycle.LiveData
import com.ustadmobile.lib.db.entities.ScrapeQueueItem
import com.ustadmobile.lib.db.entities.ScrapeQueueItemWithScrapeRun

@DoorDao
expect abstract class ScrapeQueueItemDao : BaseDao<ScrapeQueueItem> {


    @Query("SELECT * FROM ScrapeQueueItem WHERE status = $STATUS_PENDING AND itemType = :itemType ORDER BY priority ASC LIMIT 10")
    abstract fun findNextQueueItems(itemType: Int): LiveData<List<ScrapeQueueItem>>

    @Query("UPDATE ScrapeQueueItem SET status = :status, errorCode = :errorCode WHERE sqiUid = :uid")
    abstract fun updateSetStatusById(uid: Int, status: Int, errorCode: Int)

    @Query("SELECT * from ScrapeQueueItem WHERE runId = :runId AND scrapeUrl = :indexUrl LIMIT 1")
    abstract fun getExistingQueueItem(runId: Int, indexUrl: String): ScrapeQueueItem?

    @Query("SELECT * from ScrapeQueueItem WHERE runId = :runId AND sqiContentEntryUid = :entryUid LIMIT 1")
    abstract fun findExistingQueueItem(runId: Int, entryUid: Long): ScrapeQueueItem?

    @Query("UPDATE ScrapeQueueItem SET timeStarted = :timeStarted WHERE sqiUid = :uid")
    abstract fun setTimeStarted(uid: Int, timeStarted: Long)

    @Query("UPDATE ScrapeQueueItem SET timeFinished = :timeFinished WHERE sqiUid = :uid")
    abstract fun setTimeFinished(uid: Int, timeFinished: Long)

    @Query("""SELECT ScrapeQueueItem.*, ScrapeRun.* FROM ScrapeQueueItem 
                    LEFT JOIN ScrapeRun ON  ScrapeQueueItem.runId = ScrapeRun.scrapeRunUid
                    WHERE ScrapeQueueItem.sqiUid = :sqiUid""")
    abstract fun findByUid(sqiUid: Int): ScrapeQueueItemWithScrapeRun?


}
